### PR TITLE
Add lab4.

### PR DESCRIPTION
**Pull request by KonovalovaVV**

### CODE REVIEW

File: Lab4/MarkerFunction.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+			Sleep(sleep_time);
+			positions[position] = 0;
+			marked_items++;
+			
+		}
+		else
+		{
+			LeaveCriticalSection(&critical_section);
+
+			std::cout << "\nThread num " << number + 1 << "\n";
```
Comment (pashatyl) [inline Lab4/MarkerFunction.cpp]:
```suggestion
			std::cout << "\nThread num " << number + 1 << std::endl;
```

File: Lab4/MarkerFunction.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			std::cout << "Elements marked " << marked_items << "\n";
+			std::cout << "Position of unmarkable element " << position << "\n";
```
Comment (pashatyl) [inline Lab4/MarkerFunction.cpp]: Would this lines be printed as single block?

File: Lab4/MarkerFunction.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
+#include <Windows.h>
+#include <stdio.h>
+#include <math.h>
+#include <iostream>
+
+const int sleep_time = 5;
+
+void MarkerFunc(int number, HANDLE start_event, HANDLE continue_event,
```
Comment (pashatyl) [inline Lab4/MarkerFunction.cpp]:
```suggestion
void MarkerFunc(int id, HANDLE start_event, HANDLE continue_event, 
```

File: Lab4/MarkerFunction.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+			if (terminate_event_array[number])
+			{
+				EnterCriticalSection(&critical_section);
+				for (int i = 0; i < array_size; i++)
+				{
+					if (positions[i] != -1)
+					{
+						array[i] = 0;
+					}
+				}
```
Comment (pashatyl) [inline Lab4/MarkerFunction.cpp]: Could you do it faster?

File: Lab4/MarkerFunction.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	{
+		int position = rand() % array_size;
+		EnterCriticalSection(&critical_section);
+		if (marked_items < array_size && array[position] == 0)
+		{
+			Sleep(sleep_time);
+			array[position] = number + 1;
+			LeaveCriticalSection(&critical_section);
+			Sleep(sleep_time);
+			positions[position] = 0;
```
Comment (pashatyl) [inline Lab4/MarkerFunction.cpp]: Why do you need to store ints, when you use them as bool? Why 0 and -1, not 1 and 0

File: Lab4/Lab4.cpp [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	CRITICAL_SECTION critical_section;
+
+	std::cout << "Enter array size ";
+	std::cin >> array_size;
+
+	array = new int[array_size];
+	for (int i = 0; i < array_size; i++)
+	{
+		array[i] = 0;
+	}
```
Comment (pashatyl) [inline Lab4/Lab4.cpp]: std::vector or, at least, memset

File: Lab4/Lab4.cpp [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+		if (alive_thread_amount == 0)
+		{
+			std::cout << "\nNo threads are available....\n";
+			break;
+		}
+
+		std::cout << "\nPlease, enter thread number to terminate : ";
+		int number;
+		std::cin >> number;
+		terminate_event_array[number - 1] = TRUE;
```
Comment (pashatyl) [inline Lab4/Lab4.cpp]:
```suggestion
		terminate_event_array[number - 1] = true;
```

File: Lab4/Lab4.cpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+		PulseEvent(continue_event);
```
Comment (pashatyl) [inline Lab4/Lab4.cpp]: Any thoughts about using PulseEvent here?

File: Lab4/Lab4.cpp (hunk 6 above)
Comment (KonovalovaVV) [reply Lab4/Lab4.cpp:31]: Change to std::vector

File: Lab4/Lab4.cpp (hunk 8 above)
Comment (KonovalovaVV) [reply Lab4/Lab4.cpp:81]: “… unreliable and should not be used by new applications".

File: Lab4/MarkerFunction.cpp (hunk 5 above)
Comment (KonovalovaVV) [reply Lab4/MarkerFunction.cpp:32]: Ok, change to std::vector<bool>

File: Lab4/MarkerFunction.cpp (hunk 4 above)
Comment (KonovalovaVV) [reply Lab4/MarkerFunction.cpp:54]: Add a counter to compare with marked_items

**Pull request merged**